lib/file-source/src/checkpointer.rs

```diff
@@ -94,7 +148,27 @@ impl Checkpointer {
         }
     }
 
+    /// Persist the current checkpoints state to disk, making our best effort to do so in an atomic
+    /// way that allow for recovering the previous state in the event of a crash.
     pub fn write_checkpoints(&mut self) -> Result<usize, io::Error> {
+        // First write the new checkpoints to a tmp file and flush it fully to disk. If vector
+        // dies anywhere during this section, the existing stable file will still be in its current
+        // valid state and we'll be able to recover.
+        let mut f = fs::File::create(&self.tmp_file_path)?;
```

Comment:
It strikes me that we could probably open the file append only, then  write checkpoints line by line, and always just read the last valid line? I'm not sure how we'd handle trimming the size though.

Reply:
That approach seemed to raise more questions without really solving any issues, so I did not go with it.

Reply:
Works for me. :)